File: java/revenj-core/src/main/java/org/revenj/database/postgres/converters/EnumConverter.java
package org.revenj.database.postgres.converters;

import org.revenj.database.postgres.PostgresBuffer;
import org.revenj.database.postgres.PostgresWriter;
import org.revenj.database.postgres.PostgresReader;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.function.Function;

public abstract class EnumConverter {

	public static void serializeURI(PostgresBuffer sw, Enum value) {
		if (value == null) return;
		sw.addToBuffer(value.name());
	}

	public static <T extends Enum> List<T> parseCollection(
			PostgresReader reader,
			int context,
			T defaultValue,
			Function<PostgresReader, T> factory) throws IOException {
		int cur = reader.read();
		if (cur == ',' || cur == ')') {
			return null;
		}
		boolean escaped = cur != '{';
		if (escaped) {
			reader.read(context);
		}
		cur = reader.peek();
		if (cur == '}') {
			if (escaped) {
				reader.read(context);
			} else {
				reader.read(2);
			}
			return new ArrayList<>(0);
		}
		int innerContext = context == 0 ? 1 : context << 1;
		List<T> list = new ArrayList<>();
		do {
			cur = reader.read();
			if (cur == '"' || cur == '\\') {
				cur = reader.read(innerContext);
				reader.initBuffer((char) cur);
				reader.fillUntil('\\', '"');
				list.add(factory.apply(reader));
				cur = reader.read(innerContext + 1);
			} else {
				reader.initBuffer((char) cur);
				reader.fillUntil(',', '}');
				cur = reader.read();
				if (reader.bufferMatches("NULL")) {
					list.add(defaultValue);
				} else {
					list.add(factory.apply(reader));
				}
			}
		} while (cur == ',');
		if (escaped) {
			reader.read(context + 1);
		} else {
			reader.read();
		}
		return list;
	}

	public static <T extends Enum> PostgresTuple toTuple(T value) {
		if (value == null) return null;
		return new EnumTuple(value.name());
	}

	private static class EnumTuple extends PostgresTuple {
		private final String value;
		private final boolean escapeArray;

		EnumTuple(String value) {
			this.value = value;
			escapeArray = "NULL".equals(value);
		}

		public boolean mustEscapeRecord() {
			return false;
		}

		public boolean mustEscapeArray() {
			return escapeArray;
		}

		public String buildTuple(boolean quote) {
			return quote ? "'" + value + "'" : value;
		}

		public void insertRecord(PostgresWriter sw, String escaping, Mapping mappings) {
			sw.write(value);
		}

		public void insertArray(PostgresWriter sw, String escaping, Mapping mappings) {
			if (mappings != null) {
				for (int x = 0; x < value.length(); x++) {
					mappings.map(sw, value.charAt(x));
				}
			} else sw.write(value);
		}
	}
}

File: csharp/Core/Revenj.Core/DatabasePersistence/Postgres/Converters/EnumConverter.cs
using System;
using System.Collections.Generic;
using System.IO;
using Revenj.Utility;

namespace Revenj.DatabasePersistence.Postgres.Converters
{
	public static class EnumConverter
	{
		public static List<T?> ParseNullableCollection<T>(BufferedTextReader reader, int context, Func<BufferedTextReader, T> factory)
			where T : struct
		{
			var cur = reader.Read();
			if (cur == ',' || cur == ')')
				return null;
			var espaced = cur != '{';
			if (espaced)
				reader.Read(context);
			cur = reader.Peek();
			if (cur == '}')
			{
				if (espaced)
					reader.Read(context);
				else
					reader.Read(2);
				return new List<T?>(0);
			}
			var innerContext = context == 0 ? 1 : context << 1;
			var list = new List<T?>();
			do
			{
				cur = reader.Read();
				if (cur == '"' || cur == '\\')
				{
					cur = reader.Read(innerContext);
					reader.InitBuffer((char)cur);
					reader.FillUntil('\\', '"');
					list.Add(factory(reader));
					cur = reader.Read(innerContext + 1);
				}
				else
				{
					reader.InitBuffer((char)cur);
					reader.FillUntil(',', '}');
					cur = reader.Read();
					if (reader.BufferMatches("NULL"))
						list.Add(null);
					else
						list.Add(factory(reader));
				}
			} while (cur == ',');
			if (espaced)
				reader.Read(context + 1);
			else
				reader.Read();
			return list;
		}

		public static List<T> ParseCollection<T>(BufferedTextReader reader, int context, Func<BufferedTextReader, T> factory)
			where T : struct
		{
			var cur = reader.Read();
			if (cur == ',' || cur == ')')
				return null;
			var espaced = cur != '{';
			if (espaced)
				reader.Read(context);
			cur = reader.Peek();
			if (cur == '}')
			{
				if (espaced)
					reader.Read(context);
				else
					reader.Read(2);
				return new List<T>(0);
			}
			var innerContext = context == 0 ? 1 : context << 1;
			var list = new List<T>();
			do
			{
				cur = reader.Read();
				if (cur == '"' || cur == '\\')
				{
					cur = reader.Read(innerContext);
					reader.InitBuffer((char)cur);
					reader.FillUntil('\\', '"');
					list.Add(factory(reader));
					cur = reader.Read(innerContext + 1);
				}
				else
				{
					reader.InitBuffer((char)cur);
					reader.FillUntil(',', '}');
					cur = reader.Read();
					if (reader.BufferMatches("NULL"))
						list.Add(default(T));
					else
						list.Add(factory(reader));
				}
			} while (cur == ',');
			if (espaced)
				reader.Read(context + 1);
			else
				reader.Read();
			return list;
		}

		public static IPostgresTuple ToTuple<T>(T value)
			where T : struct
		{
			return new EnumTuple(value.ToString());
		}

		public static IPostgresTuple ToTuple<T>(T? value)
			where T : struct
		{
			return value != null ? new EnumTuple(value.ToString()) : null;
		}

		class EnumTuple : IPostgresTuple
		{
			private readonly string Value;
			private readonly bool EscapeArray;

			public EnumTuple(string value)
			{
				this.Value = value;
				EscapeArray = value == "NULL";
			}

			public bool MustEscapeRecord { get { return false; } }
			public bool MustEscapeArray { get { return EscapeArray; } }

			public string BuildTuple(bool quote)
			{
				return quote ? "'" + Value + "'" : Value;
			}

			private void Escape(TextWriter sw, string escaping, Action<TextWriter, char> mappings)
			{
				if (mappings != null)
				{
					foreach (var c in Value)
						mappings(sw, c);
				}
				else sw.Write(Value);
			}

			public void InsertRecord(TextWriter sw, char[] buf, string escaping, Action<TextWriter, char> mappings)
			{
				sw.Write(Value);
			}

			public void InsertArray(TextWriter sw, char[] buf, string escaping, Action<TextWriter, char> mappings)
			{
				if (mappings != null)
					foreach (var c in Value)
						mappings(sw, c);
				else
					sw.Write(Value);
			}
		}
	}
}
